File: src/Coding_Interview_Patterns_ByteByteGo/ch1_two_pointers/python3/largest_container.py
from typing import List


'''
Approach: Two Pointers (Inward Traversal)
Time Complexity: O(n)
Space Complexity: O(1)
'''
def largest_container(heights: List[int]) -> int:
    max_water = 0
    left = 0
    right = len(heights) - 1
    while left < right:
        water = min(heights[left], heights[right]) * (right - left)
        max_water = max(max_water, water)
        if heights[left] < heights[right]:
            left += 1
        elif heights[left] > heights[right]:
            right -= 1
        else:
            left += 1
            right -= 1
    return max_water


# '''
# Approach: Brute Force
# Time Complexity: O(n^2)
# Space Complexity: O(1)
# '''
# def largest_container(heights: List[int]) -> int:
#     max_water = 0
#     n = len(heights)
#     for i in range(n):
#         for j in range(i + 1, n):
#             water = min(heights[i], heights[j]) * (j - i)
#             max_water = max(max_water, water)
#     return max_water

File: src/ByteByteGo/Coding_Interview_Patterns/ch01_two_pointers/LargestContainer.java
//package Coding_Interview_Patterns_ByteByteGo.ch01_two_pointers;
//
//import static org.junit.jupiter.api.Assertions.*;
//
///**
// * <a href="https://bytebytego.com/exercises/coding-patterns/two-pointers/largest-container">Largest Container</a>
// */
//public class LargestContainer {
//
//  public static void main(String[] args) {
//    assertEquals(24, largestContainer(new int[] {2, 7, 8, 3, 7, 6}));
//    assertEquals(0, largestContainer(new int[] {}));
//    assertEquals(0, largestContainer(new int[] {1}));
//    assertEquals(0, largestContainer(new int[] {0, 1, 0}));
//    assertEquals(9, largestContainer(new int[] {3, 3, 3, 3}));
//    assertEquals(2, largestContainer(new int[] {1, 2, 3}));
//    assertEquals(2, largestContainer(new int[] {3, 2, 1}));
//  }
//
//  /**
//   * Approach: Two Pointers (Inward Traversal) <br>
//   * Time Complexity: O(n) <br>
//   * Space Complexity: O(1) <br>
//   */
//  public static int largestContainer(int[] heights) {
//    int maxWater = 0;
//    int left = 0;
//    int right = heights.length - 1;
//    while (left < right) {
//      int water = Math.min(heights[left], heights[right]) * (right - left);
//      maxWater = Math.max(maxWater, water);
//      if (heights[left] < heights[right]) {
//        left++;
//      } else if (heights[left] > heights[right]) {
//        right--;
//      } else {
//        left++;
//        right--;
//      }
//    }
//    return maxWater;
//  }
//
////  /**
////   * Approach: Brute Force <br>
////   * Time Complexity: O(n^2) <br>
////   * Space Complexity: O(1) <br>
////   */
////  public static int largestContainer(int[] heights) {
////    int maxWater = 0;
////    int n = heights.length;
////    for (int i = 0; i < n; i++) {
////      for (int j = i + 1; j < n; j++) {
////        int water = Math.min(heights[i], heights[j]) * (j - i);
////        maxWater = Math.max(maxWater, water);
////      }
////    }
////    return maxWater;
////  }
//}
